File: sources/fortran/11_parallel_sum.f90
! task 11 parallel_sum — expected output: 7500000075000000
! build: gfortran -O3 -fopenmp -o prog 11_parallel_sum.f90 (flang -O3 -fopenmp -o prog 11_parallel_sum.f90)    run: ./prog
!
! OpenMP is the threading route here: Fortran's standard library has none.
! The four chunks are the same work as task 02, so the two are comparable.

program main
  implicit none
  integer(kind=8) :: t, i, acc, total

  total = 0_8
  !$omp parallel do reduction(+:total) private(acc, i) schedule(static)
  do t = 0_8, 3_8
    acc = 0_8
    do i = t*25000000_8, (t + 1_8)*25000000_8 - 1_8
      select case (mod(i, 4_8))
      case (0_8)
        acc = acc + 1_8
      case (1_8)
        acc = acc + i
      case (2_8)
        acc = acc + 2_8*i
      case (3_8)
        acc = acc + 3_8*i
      end select
    end do
    total = total + acc
  end do
  !$omp end parallel do

  write(*,'(i0)') total
end program main

File: sources/r/11_parallel_sum.R
# task 11 parallel_sum — expected output: 7500000075000000
# build: Rscript 11_parallel_sum.R    run: Rscript 11_parallel_sum.R
#
# R has no threads: plain R code never runs on two threads. The bundled
# parallel package's PSOCK cluster is the mechanism the benchmark sanctions
# here, so this is four separate R processes over sockets, not four threads.
# PSOCK workers start with an empty environment, so N is exported to them.

N <- 25000000

cl <- parallel::makeCluster(4, type = "PSOCK")
parallel::clusterExport(cl, varlist = c("N"))
parts <- parallel::parSapply(cl, 0:3, function(t) {
  acc <- 0
  i <- t * N
  end <- i + N
  while (i < end) {
    acc <- acc + switch(as.integer(i %% 4) + 1, 1, i, 2 * i, 3 * i)
    i <- i + 1
  }
  acc
})
parallel::stopCluster(cl)

cat(sprintf("%.0f\n", sum(parts)))
